### PR TITLE
exclude dead branches

### PR DESCRIPTION
**Pull request by Rich-Harris**
Follow-up to https://github.com/rollup/rollup/issues/208#issuecomment-150096281. This PR excludes dead branches, wherever it's possible to determine that through static analysis.

This goes slightly further than what e.g. UglifyJS does:

```js
var obj = {};
obj.foo = function () {
  console.log( 'this function should be excluded' );
};

if ( false ) obj.foo();
```

Rollup will exclude `obj.foo` but UglifyJS won't, as far as I can tell (there may be an option that I don't know about).

Alongside https://github.com/rollup/rollup-plugin-replace, this should make it easier to make different bundles for different targets (e.g. browser vs node, dev vs prod).

### CODE REVIEW

File: src/ast/conditions.js [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+	if ( a.type === 'Literal' ) return a.value === b.value;
+	if ( a.type === 'Identifier' ) return a.name === b.name;
+
+	return false;
+}
+
+function nodesAreNotEqual ( a, b ) {
+	if ( a.type !== b.type ) return false;
+	if ( a.type === 'Literal' ) return a.value != b.value;
+	if ( a.type === 'Identifier' ) return a.name != b.name;
```
Comment (Victorystick) [inline src/ast/conditions.js]: Any reason to use `!=` here while you use `===` above?

File: src/ast/conditions.js [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+	if ( node.type === 'LogicalExpression' ) {
+		if ( node.operator === '&&' ) return isTruthy( node.left ) && isTruthy( node.right );
+		if ( node.operator === '||' ) return isTruthy( node.left ) || isTruthy( node.right );
+		return false;
+	}
+
+	return false;
+}
+
+export function isFalsy ( node ) {
```
Comment (Victorystick) [inline src/ast/conditions.js]: It seems odd to duplicate all this code.

How about something like:

```js
function _isTruthy ( node ) {
	...

	// `null` signifies undeterminable
	return null;
}

export function isTruthy ( node ) {
	// replace `null` with `false`
	return _isTruthy( node ) || false;
}

export function isFalsy ( node ) {
	const truthy = _isTruthy( node );

	// replace `null` with `false`, and invert the truthy value.
	return truthy === null ? false : !truthy;
}
```

File: src/ast/conditions.js [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+}
+
+function isNotEqualTest ( node ) {
+	return node.type === 'BinaryExpression' && ( node.operator === '!==' || node.operator === '!=' );
+}
+
+function nodesAreEqual ( a, b ) {
+	if ( a.type !== b.type ) return false;
+	if ( a.type === 'Literal' ) return a.value === b.value;
+	if ( a.type === 'Identifier' ) return a.name === b.name;
```
Comment (Victorystick) [inline src/ast/conditions.js]: This will fail for
```js
var a = NaN;

if ( a === a ) ...
```

Comment (Rich-Harris) [inline src/ast/conditions.js]: good catch. fixed locally

File: src/ast/conditions.js (hunk 1 above)
Comment (Rich-Harris) [inline src/ast/conditions.js]: yeah, it's to do with strict vs non-strict equality. though now that you mention it, i think it'd be better to have separate `nodesAreStrictEqual` and `nodesAreNotStrictEqual` functions

File: src/ast/conditions.js (hunk 2 above)
Comment (Rich-Harris) [inline src/ast/conditions.js]: :+1: that makes a lot more sense

File: src/ast/conditions.js (hunk 3 above)
Comment (Victorystick) [inline src/ast/conditions.js]: No, I mean; what if an _identifier_ has the _value_ `NaN`?

**Comment by Victorystick** [comment]
What I mean is, what if an _identifier_ has the _value_ `NaN`?

The assumption `a == a -> true` is invalid.

**Comment by Rich-Harris** [comment]
> What I mean is, what if an identifier has the value NaN?

Ah. Good point. Bloody NaN! Is there any more useless feature in any programming language?

So do we skip identifier equality checks altogether? I guess the only alternative is to have an option controlling whether or not to care about NaN, which is territory I'd rather not get into

**Comment by Victorystick** [comment]
I fear the best thing to do is to _skip-identifiers_. It's unlikely to end up with `NaN`s like that, but it could happen. Unless we can be sure of the identifiers' values.

Say our `Declaration`s had a `value` property, then we could compare actual values. `const` variables would be easy. It'd be a best-effort thing regardless.

**Comment by Rich-Harris** [comment]
agreed

**Comment by Victorystick** [comment]
@Rich-Harris Mind if I go ahead and merge this?

**Comment by Rich-Harris** [comment]
Yep let's do it. We should probably merge #253 too and release it as 0.21 (so that people can pin 0.20 if it turns out the algorithm fails in certain scenarios)

**Pull request merged**